care/facility/migrations/0466_camera_presets.py

```diff
@@ -0,0 +1,171 @@
+# Generated by Django 4.2.8 on 2024-05-30 06:56
+
+import uuid
+
+import django.db.models.deletion
+from django.conf import settings
+from django.core.paginator import Paginator
+from django.db import migrations, models
+from django.db.models import F, Window
+from django.db.models.functions import RowNumber
+
+import care.utils.models.validators
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+        ("facility", "0465_merge_20240923_1045"),
+    ]
+
+    def delete_asset_beds_without_asset_class(apps, schema_editor):
+        AssetBed = apps.get_model("facility", "AssetBed")
+        AssetBed.objects.filter(asset__asset_class__isnull=True).delete()
+
+    def backfill_camera_presets(apps, schema_editor):
+        AssetBed = apps.get_model("facility", "AssetBed")
+        CameraPreset = apps.get_model("facility", "CameraPreset")
+
+        paginator = Paginator(
+            AssetBed.objects.annotate(
+                row_number=Window(
+                    expression=RowNumber(),
+                    partition_by=[F("asset"), F("bed")],
+                    order_by=F("id").asc(),
+                )
+            )
+            .filter(deleted=False, asset__asset_class="ONVIF")
+            .order_by("asset", "bed", "id"),
+            1000,
+        )
+
+        for page_number in paginator.page_range:
+            assetbeds_to_delete = []
+            presets_to_create = []
+
+            for asset_bed in paginator.page(page_number).object_list:
+                name = asset_bed.meta.get("preset_name")
+
+                if position := asset_bed.meta.get("position"):
+                    presets_to_create.append(
+                        CameraPreset(
+                            name=name,
+                            asset_bed=AssetBed.objects.filter(
+                                asset=asset_bed.asset, bed=asset_bed.bed
+                            ).order_by("id")[0],
+                            position={
+                                "x": position["x"],
+                                "y": position["y"],
+                                "zoom": position["zoom"],
+                            },
+                            is_migrated=True,
+                        )
+                    )
+                    if asset_bed.row_number != 1:
+                        assetbeds_to_delete.append(asset_bed.id)
+                else:
+                    assetbeds_to_delete.append(asset_bed.id)
+
+            CameraPreset.objects.bulk_create(presets_to_create)
+            AssetBed.objects.filter(id__in=assetbeds_to_delete).update(
+                deleted=True, meta={}
+            )
+        AssetBed.objects.filter(deleted=False, asset__asset_class="ONVIF").update(
```

Comment:
Will remove it from this PR and create separate PR to remove it.

Reply:
The duplicate asset-bed records would have to be marked as deleted for the unique constraint to be applied, and for the front-end to know which asset-bed record to link to when the preset is being created. 

We can however skip deleting the meta data from these old records so that data is still preserved, just that duplicate asset-bed link records would no longer be accessible by the viewset.